File: polynomial_linear_regression/polynomial_regression.py
# Polynomial Regression

# Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Importing the dataset
dataset = pd.read_csv('polynomial_regression_Position_Salaries.csv')
X = dataset.iloc[:, 1:2].values # always make sure that independent variables are in matrix that is why 1:2
y = dataset.iloc[:, 2].values   #and Dependent varible is in vector

# Splitting the dataset into the Training set and Test set   # no need to split dataset in training set test set bcz we have only 10 observations here
'''from sklearn.cross_validation import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)'''

# Feature Scaling
"""from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
X_train = sc_X.fit_transform(X_train)
X_test = sc_X.transform(X_test)
sc_y = StandardScaler()
y_train = sc_y.fit_transform(y_train)"""

#Fitting Linear Regression to dataset      # this regressor is created to compare the result between linear regression model and Polynomial Regression
from sklearn.linear_model import LinearRegression
lin_reg = LinearRegression()
lin_reg.fit(X,y)
 
# Fitting Polynomial Regression to the dataset
from sklearn.preprocessing import PolynomialFeatures
poly_reg = PolynomialFeatures(degree = 4)    #this poly_reg object is a transformer tool that is going to transfer matrix of X into new matrix X  which will contain columns of X,X^2,X^3..... here as we used degree = 2 it will create X and X^2 this object will also contain one more column which will have only 1s which will take care of that constant variable in expression
X_poly = poly_reg.fit_transform(X)      
    #now we will fit this X_poly matrix into the linear regression model 
lin_reg_2 = LinearRegression()
lin_reg_2.fit(X_poly,y)       


# Visualising the linear Regression results 
plt.scatter(X,y,color = 'red')
plt.plot(X, lin_reg.predict(X),color = 'blue')
plt.title('salary detector (Linear regression)')
plt.xlabel('Position level')
plt.ylabel('Salary')
plt.show()                   # this graph is not at all showing best fit for the data the predicted results were too far


# Visualising the Polynomial Regression results
X_grid = np.arange(min(X), max(X), 0.1) # this will contain all the levels (10 levels according to dataset) plus incremental steps between the levels with a resolution of 0.1  
X_grid = X_grid.reshape((len(X_grid),1)) # to convert vector to matrix
plt.scatter(X,y,color = 'red')
plt.plot(X_grid, lin_reg_2.predict(poly_reg.fit_transform(X_grid)),color = 'blue')
plt.title('salary detector (Polinomial regression)')
plt.xlabel('Position level')
plt.ylabel('Salary')
plt.show()                              # polinomial regression model fits much better to our dataset problem than linear regression model

        
# Predicting new result with linear regression 
lin_reg.predict(6.5)        
        
# Predicting new result with polinomial regression
lin_reg_2.predict(poly_reg.fit_transform(6.5))        
        

File: polynomial_linear_regression/polynomial_regression.R
# Polynomial Regression

# Importing the dataset
dataset = read.csv('polynomial_regression_Position_Salaries.csv')
dataset = dataset[2:3] #index in r starts from '1'

# Splitting the dataset into the Training set and Test set  # as the data has only 10 observation we don not need to split data in training set and test set 
# install.packages('caTools')
# library(caTools)
# set.seed(123)
# split = sample.split(dataset$DependentVariable, SplitRatio = 0.8)
# training_set = subset(dataset, split == TRUE)
# test_set = subset(dataset, split == FALSE)

# Feature Scaling   # feature scaling is also not required here 
# training_set = scale(training_set)
# test_set = scale(test_set)

# fitting Liner regression to the dataset 
lin_reg = lm(formula = Salary ~ Level,
             data = dataset)

# Fitting Polynonmial Regression to dataset
  # here we are going to add polynomial Features, they additional independent variables which are in this case 'Level^2 , Level^3, Level^4 .....'
  # in short what we can conclude from this is polynomial regression module is nothing but Multiple Linear regression model that is composed of one independent variable and additional independent variables that are the polynomial terms of the first independent variable
dataset$Level2 = dataset$Level^2  # this is a polynomial feature or we can say we are adding one extra column here which is the power 2 of Level
dataset$Level3 = dataset$Level^3
dataset$Level4 = dataset$Level^4
poly_reg = lm(formula = Salary ~ . ,
              data = dataset)

#Visualizing the linear regression Results
ggplot()+
  geom_point(aes(x= dataset$Level , y = dataset$Salary),
             color = 'red')+
  geom_line(aes(x= dataset$Level , y = predict(lin_reg,newdata = dataset)),
            color = 'blue')+
  ggtitle('truth or bluff (Linear Regression)')+
  xlab('Level')+
  ylab('Salary')
  
#Visualizing the polynomial Regression results
ggplot()+
  geom_point(aes(x= dataset$Level , y = dataset$Salary),
             color = 'red')+
  geom_line(aes(x= dataset$Level , y = predict(poly_reg,newdata = dataset)),
            color = 'blue')+
  ggtitle('truth or bluff (Polynomial Regression)')+
  xlab('Level')+
  ylab('Salary')


# predicting new result with Linear Regression
y_pred = predict(lin_reg,data.frame(Level = 6.5)) #here we have to create new dataset for single data that we want to predict by using data.dataframe we can add extra cell in our data to add specific value in it
# here data.frame(Level = 6.5)  actually created dataset with Level variable and 6.5 entry in Level column


# predicting new result with Polynomial Regression

y_pred = predict(poly_reg,data.frame(Level = 6.5,
                                    Level2 = 6.5^2,
                                    Level3 = 6.5^3,
                                    Level4 = 6.5^4))



# Visualising the Regression Model results (for higher resolution and smoother curve)
# install.packages('ggplot2')
x_grid = seq(min(dataset$Level), max(dataset$Level), 0.1)  
ggplot() +
  geom_point(aes(x = dataset$Level, y = dataset$Salary),
             colour = 'red') +
  geom_line(aes(x = x_grid, y = predict(poly_reg,
                                        newdata = data.frame(Level = x_grid,
                                                             Level2 = x_grid^2,
                                                             Level3 = x_grid^3,
                                                             Level4 = x_grid^4))),
            colour = 'blue') +
  ggtitle('Truth or Bluff (Polynomial Regression)') +
  xlab('Level') +
  ylab('Salary')


